### PR TITLE
i#5623: AARCH64 feature FEAT_PAUTH2 is now recognized

### PR DESCRIPTION
**Pull request by philramsey-arm**
Enhance the AARCH64 feature interface so that we can correctly identify features identifed by more than one register nibble.

Update the identification of FEATURE_PAUTH so that all six identification nibbles are checked.

Remove the automatic assumption that if the value of the nibble in the register is greater than the expected value then the feature is implemented.

Issue: #5623

### CODE REVIEW

File: core/arch/proc_api.h [hunk 1]
```diff
@@ -312,59 +311,69 @@ typedef enum {
...
-    FEATURE_F64MM = DEF_FEAT(AA64ZFR0, 14, 1, 0),   /**< SVE FP64 matrix multiplication */
-    FEATURE_SVE2 = DEF_FEAT(AA64ZFR0, 0, 1, 0),     /**< Scalable vectors 2 (AArch64) */
-    FEATURE_SVEAES = DEF_FEAT(AA64ZFR0, 1, 1, 0),   /**< SVE2 + AES(AArch64) */
-    FEATURE_SVESHA3 = DEF_FEAT(AA64ZFR0, 8, 1, 0),  /**< SVE2 + SHA3(AArch64) */
-    FEATURE_SVESM4 = DEF_FEAT(AA64ZFR0, 10, 1, 0),  /**< SVE2 + SM4(AArch64) */
-    FEATURE_SVEBitPerm = DEF_FEAT(AA64ZFR0, 4, 1, 0), /**< SVE2 + BitPerm(AArch64) */
-    FEATURE_MTE = DEF_FEAT(AA64PFR1, 2, 1, 0),        /**< Memory Tagging Extension */
-    FEATURE_BTI = DEF_FEAT(AA64PFR1, 0, 1, 0),        /**< Branch Target Identification*/
+    /*                      Register NibblePos FeatVal Flags */
+    FEATURE_AESX = DEF_FEAT(AA64ISAR0, 1, 1, FEAT_GREATER_EQUAL),
```
Comment (AssadHashmi) [inline core/arch/proc_api.h:329]: These should not lose their Doxygen commenting, e.g. `/**< AES<x> (AArch64) */`.

**Review by AssadHashmi — CHANGES_REQUESTED** [review]

File: core/arch/proc_api.h (hunk 1 above)
Comment (philramsey-arm) [reply core/arch/proc_api.h]: Done

File: core/arch/aarch64/proc.c [hunk 2]
```diff
@@ -293,71 +265,97 @@ enable_all_test_cpu_features()
...
     dr_set_sve_vector_length(256);
 }
 
+uint64
+get_reg_val(feature_reg_idx_t feat_reg)
+{
+    return cpu_info.features.isa_features[feat_reg];
+}
+
 bool
```
Comment (derekbruening) [inline core/arch/aarch64/proc.c:282]: Make all non-global functions static

File: core/arch/proc_shared.c [hunk 3]
```diff
@@ -85,11 +85,10 @@ cpu_info_t cpu_info = {
     CACHE_SIZE_UNKNOWN,
     CACHE_SIZE_UNKNOWN,
     CACHE_SIZE_UNKNOWN,
-#if defined(RISCV64)
-    /* FIXME i#3544: Not implemented */
```
Comment (derekbruening) [inline core/arch/proc_shared.c:89]: This comment was lost.

File: suite/tests/client-interface/features.dll.c [hunk 4]
```diff
@@ -71,12 +73,36 @@ read_hw_features(char *feat_str)
...
+check_for_pauth()
+{
+    uint64 id_aa64isar1_el1 = 0;
+
+    asm("mrs %0, ID_AA64ISAR1_EL1" : "=r"(id_aa64isar1_el1));
+
+    ushort gpi = (id_aa64isar1_el1 >> 28) & 0xF;
+    ushort gpa = (id_aa64isar1_el1 >> 24) & 0xF;
+    ushort api = (id_aa64isar1_el1 >> 8) & 0xF;
+    ushort apa = (id_aa64isar1_el1 >> 4) & 0xF;
```
Comment (derekbruening) [inline suite/tests/client-interface/features.dll.c:91]: Line 77 says "six different nibbles": here we see 4 though.

File: suite/tests/client-interface/features.dll.c [hunk 5]
```diff
@@ -71,12 +73,36 @@ read_hw_features(char *feat_str)
...
+ * that the other nibbles are being checked correctly.
+ */
+void
+check_for_pauth()
+{
+    uint64 id_aa64isar1_el1 = 0;
+
+    asm("mrs %0, ID_AA64ISAR1_EL1" : "=r"(id_aa64isar1_el1));
+
+    ushort gpi = (id_aa64isar1_el1 >> 28) & 0xF;
```
Comment (derekbruening) [inline suite/tests/client-interface/features.dll.c:88]: Is there a more descriptive name than "gpi" (ditto for gpa, api, apa).  This is hard to understand: where did the 28,24,8,4 come from?  How about an inline comment `28 /* FEAT_FOO*/` or named constant (is DR already exported a named constant?  Or make a new local one.)

File: core/arch/proc_api.h [hunk 6]
```diff
@@ -166,28 +166,27 @@ typedef struct {
...
  */
 typedef struct {
-    uint64 flags_aa64isar0; /**< AArch64 feature flags stored in ID_AA64ISAR0_EL1 */
-    uint64 flags_aa64isar1; /**< AArch64 feature flags stored in ID_AA64ISAR1_EL1 */
-    uint64 flags_aa64pfr0;  /**< AArch64 feature flags stored in ID_AA64PFR0_EL1 */
-    uint64 flags_aa64mmfr1; /**< AArch64 feature flags stored in ID_AA64MMFR1_EL1 */
-    uint64 flags_aa64dfr0;  /**< AArch64 feature flags stored in ID_AA64DFR0_EL1 */
-    uint64 flags_aa64zfr0;  /**< AArch64 feature flags stored in ID_AA64ZFR0_EL1 */
-    uint64 flags_aa64pfr1;  /**< AArch64 feature flags stored in ID_AA64PFR1_EL1 */
+    uint64 isa_features[AA64_NUM_FEAT_REGS];
```
Comment (derekbruening) [inline core/arch/proc_api.h:187]: A public type needs a Doxygen comment for every field.

File: core/arch/proc_api.h [hunk 7]
```diff
@@ -166,28 +166,27 @@ typedef struct {
  * document both defines, for X86 and for AARCHXX. See also i#5496.
  */
 #ifdef AARCHXX
+
+typedef enum {
+    AA64ISAR0,
```
Comment (derekbruening) [inline core/arch/proc_api.h:171]: A public type needs a Doxygen comment for each member.

File: core/arch/aarch64/proc.c [hunk 8]
```diff
@@ -293,71 +265,97 @@ enable_all_test_cpu_features()
...
+
+bool
+proc_has_feature(feature_bit_t feature_bit)
+{
+#ifdef DR_HOST_NOT_TARGET
+    return false;
+#else
+    bool has_feature = proc_has_feature_imp(feature_bit);
+
+    /* Some features are identified in more than one place */
```
Comment (derekbruening) [inline core/arch/aarch64/proc.c:354]: nit: Missing trailing .

File: core/arch/aarch64/proc.c [hunk 9]
```diff
@@ -293,71 +265,97 @@ enable_all_test_cpu_features()
...
+
+    return has_feature;
+}
+
+/*
+ * Some features are identified by more than one nibble.
+ * In this case we need to add extra mappings between features and nibble values
+ */
+static uint32 feature_ids[] = {
+    (FEATURE_PAUTH << 16) | DEF_FEAT(AA64ISAR1, 1, 1, FEAT_GR_EQ), /* APA (QARMA5) */
```
Comment (derekbruening) [inline core/arch/aarch64/proc.c:313]: I don't understand: which feature is this?  FEATURE_PAUTH?  What is "APA (QARMA5)"?  Is it that FEATURE_PAUTH really means any one of 5 different sub-features?  Reword to `FEATURE_PAUTH.APA_QARMA5` or sthg to make it more clear?  Why aren't the FEATURE_PAUTH all grouped together in this array?

File: core/arch/aarch64/proc.c [hunk 10]
```diff
@@ -187,98 +184,73 @@ proc_init_arch(void)
...
+        LOG(GLOBAL, LOG_TOP, 1, "Processor features:\n ID_AA64ISAR2_EL1 = 0x%016lx\n",
+            cpu_info.features.isa_features[AA64ISAR2]);
+        LOG_FEATURE(FEATURE_PAUTH2);
     });
 #    endif
 #endif
 }
 
-#define GET_FEAT_REG(FEATURE) (feature_reg_idx_t)((((ushort)FEATURE) & 0x7F00) >> 8)
+#define GET_FEAT_REG(FEATURE) (feature_reg_idx_t)((((ushort)FEATURE) & 0x3F00) >> 8)
```
Comment (derekbruening) [inline core/arch/aarch64/proc.c:229]: Where is this 0x3F00 coming from?  Is this from DEF_FEAT?  How about a comment showing the bit #s and what each one is.

File: core/arch/proc_api.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (philramsey-arm) [reply core/arch/proc_api.h:171]: In the Arm ARM the register is called ID_AA64ISAR0_EL1 but I can add underscores to make it easier to read.

File: core/arch/aarch64/proc.c (hunk 2 above)
Comment (philramsey-arm) [reply core/arch/aarch64/proc.c:282]: Will do

File: core/arch/proc_shared.c (hunk 3 above)
Comment (philramsey-arm) [reply core/arch/proc_shared.c:89]: Fixed

File: core/arch/proc_api.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (AssadHashmi) [reply core/arch/proc_api.h:171]: I think it's best to leave it as-is, remaining consistent with the spec.

File: suite/tests/client-interface/features.dll.c (hunk 4 above)
Comment (philramsey-arm) [reply suite/tests/client-interface/features.dll.c:91]: Will update this to read the ID_AA64ISAR2_EL1 register as well

File: suite/tests/client-interface/features.dll.c (hunk 5 above)
Comment (philramsey-arm) [reply suite/tests/client-interface/features.dll.c:88]: I have added descriptions to each flag.

File: core/arch/aarch64/proc.c (hunk 9 above)
Comment (philramsey-arm) [reply core/arch/aarch64/proc.c:313]: I have expanded the comments and grouped the rows by feature to hopefully make this clearer.

The Arm ARM states "If the value of one or more of the ID_AA64ISAR1_EL1.{GPI, GPA, API, APA} or ID_AA64ISAR2_EL1.{GPA3, APA3} fields is nonzero, then pointer authentication is implemented."

So as well as the nibble in the definition of FEATURE_PAUTH, we have to add the 5 more in this table.

File: core/arch/aarch64/proc.c (hunk 10 above)
Comment (philramsey-arm) [reply core/arch/aarch64/proc.c:229]: I have added a breakdown of the bits in proc_api.h

File: core/arch/proc_api.h [hunk 11]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
-    AA64ZFR0,
-    AA64PFR1,
-    AA64ISAR2,
-    AA64_NUM_FEAT_REGS
-} feature_reg_idx_t;
-
 /**
- * For AArch64 this struct holds features registers' values read by MRS instructions.
- * Used by proc_get_all_feature_bits().
+ * For AArch64 the isa_features array holds features registers' values read by MRS
```
Comment (derekbruening) [inline core/arch/proc_api.h:171]: isa_features looks like an internal non-public variable so it should not be discussed in a public user-facing Doxygen comment.

File: core/arch/proc_api.h [hunk 12]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
+ * instructions. Used by proc_has_feature().
  */
+typedef enum {
+    AA64ISAR0,         /**< ID_AA64ISAR0_EL1 */
```
Comment (derekbruening) [inline core/arch/proc_api.h:175]: nit: Punctuate all Doxygen comments: so all of these here and below need a trailing `.`

File: core/arch/proc_api.h [hunk 13]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
+    AA64ISAR1,         /**< ID_AA64ISAR1_EL1 */
+    AA64PFR0,          /**< ID_AA64PFR0_EL1 */
+    AA64MMFR1,         /**< ID_AA64MMFR1_EL1 */
+    AA64DFR0,          /**< ID_AA64DFR0_EL1 */
+    AA64ZFR0,          /**< ID_AA64ZFR0_EL1 */
+    AA64PFR1,          /**< ID_AA64PFR1_EL1, */
+    AA64ISAR2,         /**< ID_AA64ISAR2_EL1 */
+    AA64_NUM_FEAT_REGS /**< Number of feature registers */
+} feature_reg_idx_t; /**< List of AArch64 registers that we read to identify features. */
```
Comment (derekbruening) [inline core/arch/proc_api.h:184]: This is colliding with the pre-comment above: not even sure which one wins (or does Doxygen combine them??) but we don't want to have this situation.  Move to the pre-comment.

File: core/arch/proc_api.h [hunk 14]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
-    AA64PFR1,
-    AA64ISAR2,
-    AA64_NUM_FEAT_REGS
-} feature_reg_idx_t;
-
 /**
- * For AArch64 this struct holds features registers' values read by MRS instructions.
- * Used by proc_get_all_feature_bits().
+ * For AArch64 the isa_features array holds features registers' values read by MRS
+ * instructions. Used by proc_has_feature().
```
Comment (derekbruening) [inline core/arch/proc_api.h:172]: I don't see this used by proc_has_feature().  In fact I don't see this used in any user-facing interface: should this be public at all?  Probably this should be moved to a non-public header.

File: core/arch/proc_api.h (hunk 12 above)
Comment (derekbruening) [inline core/arch/proc_api.h:175]: `ID_AA64ISAR0_EL1 attribute register.` would be clearer.

File: core/arch/proc_api.h [hunk 15]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
+    AA64ZFR0,          /**< ID_AA64ZFR0_EL1 */
+    AA64PFR1,          /**< ID_AA64PFR1_EL1, */
+    AA64ISAR2,         /**< ID_AA64ISAR2_EL1 */
+    AA64_NUM_FEAT_REGS /**< Number of feature registers */
+} feature_reg_idx_t; /**< List of AArch64 registers that we read to identify features. */
+
 typedef struct {
-    uint64 isa_features[AA64_NUM_FEAT_REGS];
+    uint64 isa_features[AA64_NUM_FEAT_REGS]; /**< array of ints representing the AArch64
+                                                attribute registers */
```
Comment (derekbruening) [inline core/arch/proc_api.h:188]: Use a pre-comment if it doesn't fit on one line; and always capitalize and punctuate user-facing docs.

File: core/arch/proc_api.h [hunk 16]
```diff
@@ -167,24 +167,25 @@ typedef struct {
...
+    AA64PFR0,          /**< ID_AA64PFR0_EL1 */
+    AA64MMFR1,         /**< ID_AA64MMFR1_EL1 */
+    AA64DFR0,          /**< ID_AA64DFR0_EL1 */
+    AA64ZFR0,          /**< ID_AA64ZFR0_EL1 */
+    AA64PFR1,          /**< ID_AA64PFR1_EL1, */
+    AA64ISAR2,         /**< ID_AA64ISAR2_EL1 */
+    AA64_NUM_FEAT_REGS /**< Number of feature registers */
+} feature_reg_idx_t; /**< List of AArch64 registers that we read to identify features. */
+
 typedef struct {
```
Comment (derekbruening) [inline core/arch/proc_api.h:186]: All public types need top-level Doxygen comments.

File: core/arch/proc_api.h [hunk 17]
```diff
@@ -312,8 +313,14 @@ typedef enum {
  *   supported at all but in some cases 15 means a feature is not supported at
  *   all, NSFLAG.
  * - If FEAT_EQ is set then the nibble value should match exactly, otherwise reg
- * values greater than or equal to the feature nibble are considered a match. The helper
- * macro below packs feature data into 16 bits (ushort).
+ *   values greater than or equal to the feature nibble are considered a match. The helper
+ *   macro below packs feature data into 16 bits (ushort).
+ *
+ * Bit 15: Set if a nibble value of 15 indicates feature not set
```
Comment (derekbruening) [inline core/arch/proc_api.h:319]: nit: Add trailing `.` here and lines below.  Even though not a Doxygen comment that will end up as html docs, this is still a public header presented to the user.

File: core/arch/proc_api.h (hunk 14 above)
Comment (philramsey-arm) [reply core/arch/proc_api.h:172]: The AArch64 version of proc_has_feature() calls proc_has_feature_imp() which calls get_reg_val() which reads cpu_info.features.isa_features

I guess I was copying the style of the X86 comment above.

File: core/arch/proc_api.h [hunk 18]
```diff
@@ -168,24 +168,28 @@ typedef struct {
...
+} feature_reg_idx_t;
 
+/**
+ * For AArch64 this struct contains an array of ints which cache the values of AArch64
+ * attribute registers.
+ */
 typedef struct {
-    uint64 isa_features[AA64_NUM_FEAT_REGS]; /**< array of ints representing the AArch64
-                                                attribute registers */
+    /** Array of ints representing attribute registers */
```
Comment (derekbruening) [inline core/arch/proc_api.h:191]: nit: Missing trailing `.`

**Review by derekbruening — APPROVED** [review]

**Review by AssadHashmi — APPROVED** [review]

**Pull request merged**